### PR TITLE
Add signup endpoint

### PR DESCRIPTION
**Pull request by Mukalawit**
**What does this PR do?**
This PR enables users to sign up to the PostIt Application.

**Description of the task to be completed**
For intending users to be able to use the application they must registered users,  this PR enables users to sign up, using username, a password and an email.

**Why do we need this feature?**
So that only registered users can use the application

**How can this feature be manually tested?**
Using **Postman** POST:` http:localhost:3000/api/user/signup`

Post a JSON object in the form
```
{
    "username:"username",
    "password":"password",
    "email":"email"

}
```
to sign up a user

**Testing** 
1. Create database `tests` and run this command in your terminal
```
cd server

npm run migrate:postgres

```

2. create a .env.test like this

```
DATABASE_URL="postgresql://postgres:postgres@localhost:5432/tests?schema=public"

ACCESS_TOKEN_SECRET="bb7f863dfc8b648f0c1fef9b2c245f2330d9d49535c878da6b83f79232348663e9db04254bb553891c308650dbb836bd083d813d1f5cbf0af4742659944aec9e"

REFRESH_TOKEN_SECRET="746da95529ff21d26a476eb02feb7919391967043d7a8c3315d0ae789f4019ecbd036959b60be1ef24fd681b2fbdb818d2c230bb2adf9198149ac3826b0db67e"

```

3. To test

```
npm run test

```

### CODE REVIEW

**Comment by chadwalt** [comment]
Hello @Mukalawit, the description for the MR is not clear enough, for example, I don't know how to run the application, or which endpoint to call.

Give a clear description for example detailing the following

- Description about the PR, what the PR is about, why we need the feature etc
- How to run the PR
- How to test the PR, eg. which params I need to provide as input for the endpoint

File: server/index.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+const bodyParser = require('body-parser');
+const bcrypt = require('bcrypt');
+
+const prisma = new PrismaClient();
+const saltRounds = 10;
+const app = express();
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(express.json());
+
+app.post('/api/user/signup', async (req, res) => {
```
Comment (chadwalt) [inline server/index.js]: This route wasn't tested. It is always good to test using the TDD approach. Try to look into Jest

Comment (Mukalawit) [reply server/index.js]: @chadwalt  Please review. I have added tests for the signup endpoint

**Comment by chadwalt** [comment]
Hello @Mukalawit, I have been looking at the way you named your branched and also the commit messages, and they don't follow the recommended good practices.

Here are some recommendations:

1. [Link to the best practices](https://github.com/conventional-changelog/commitlint#what-is-commitlint)
2. [We use this at work for guidance](https://github.com/Mukalawit/PostIt/files/10063525/Copy_of_Old_Git_Best_Practices.pdf)

**Comment by chadwalt** [comment]
The readme instructions are miss leading
Error when installing

Please update the readme instructions

**Comment by chadwalt** [comment]
Adding an example .env.example file would help someone trying to set up the API. The rule is to make it extremely easy for someone to run your code by providing all the necessary details of how to run it, test it, and install it.

File: server/__tests__/api.tests.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+const request = require('supertest');
+const { PrismaClient } = require('@prisma/client');
+
+const prisma = new PrismaClient({
```
Comment (chadwalt) [inline server/__tests__/api.tests.js]: I am not an expert on Prisma, but I don't think this is the right way to set up mock testing with Primsa. This inserts into the database which might become really slow for example if different pipelines in a CI/CD are running this.

Being looking at this [documentation](https://www.prisma.io/docs/guides/testing/unit-testing) where Prima client is mocked and tested.

File: server/__tests__/api.tests.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+  it('should fail if no email is submitted', async () => {
+    const response = await request(baseURL).post('/api/user/signup').send({
+      username: 'username',
+      password: 'password',
+    });
+    expect(response.statusCode).toBe(400);
+    expect(response.body.message).toBe('Invalid Payload');
+  });
+  it('should fail if no username is submitted', async () => {
```
Comment (chadwalt) [inline server/__tests__/api.tests.js]: This looks like a duplicate from the first test

**Comment by chadwalt** [comment]
I am curious, what is the logic behind the directory structure? Are you following any recommend options?

File: server/index.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    where: {
+      username
+    }
+  });
+
+  if (usernameExists) {
+    res.status(409).json({
+      message: 'username already in use'
+    });
+  } else {
```
Comment (chadwalt) [inline server/index.js]: Adding a `return` statement instead of an `else` would really help with readability.

File: server/index.js (hunk 1 above)
Comment (chadwalt) [reply server/index.js]: I also think defining a routes into their own file, would help with testing, clean structuring.

File: server/index.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
+const { PrismaClient } = require('@prisma/client');
+
+const express = require('express');
+const bodyParser = require('body-parser');
+const bcrypt = require('bcrypt');
```
Comment (chadwalt) [inline server/index.js]: Am curious, which metrics did you use to chose this package?

Comment (Mukalawit) [reply server/index.js]: @chadwalt Besides incorporating a salt to protect against rainbow table attacks, bcrypt is an adaptive function: over time, the iteration count can be increased to make it slower, so it remains resistant to brute-force search attacks even with increasing computation power, i found this as a resounding complement to bcrypt from various sources

File: server/__tests__/api.tests.js (hunk 3 above)
Comment (Mukalawit) [reply server/__tests__/api.tests.js]: It tests various inputs: email, username and password all of which contribute to the validity of the payload

**Pull request closed without merging**